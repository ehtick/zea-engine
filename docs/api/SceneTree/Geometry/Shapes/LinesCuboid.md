<a name="LinesCuboid"></a>

### LinesCuboid 
A class for generating a lines cuboid shape(Without faces).

**Parameters**
* **X([`NumberParameter`](api/SceneTree/Parameters/NumberParameter.md)):** Length of the line cuboid along the `X` axis
* **Y([`NumberParameter`](api/SceneTree/Parameters/NumberParameter.md)):** Length of the line cuboid along the `Y` axis
* **Z([`NumberParameter`](api/SceneTree/Parameters/NumberParameter.md)):** Length of the line cuboid along the `Z` axis
* **BaseZAtZero([`NumberParameter`](api/SceneTree/Parameters/NumberParameter.md)):** Property to start or not `Z` axis from position `0.


**Extends**: <code>[ProceduralLines](api/SceneTree/Geometry/Shapes/ProceduralLines.md)</code>  
<a name="new_LinesCuboid_new"></a>

### new LinesCuboid
Create a lines cuboid.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| x | <code>number</code> | <code>1</code> | The length of the line cuboid along the X axis. |
| y | <code>number</code> | <code>1</code> | The length of the line cuboid along the Y axis. |
| z | <code>number</code> | <code>1</code> | The length of the line cuboid along the Z axis. |
| baseZAtZero | <code>boolean</code> | <code>false</code> | The baseZAtZero value. |



### [Class Tests](api/SceneTree/Geometry/Shapes/LinesCuboid.test)