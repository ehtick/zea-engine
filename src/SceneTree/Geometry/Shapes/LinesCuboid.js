import { Lines } from '../Lines.js';
import {
  NumberParameter
} from '../../Parameters/NumberParameter.js';
import {
  sgFactory
} from '../../SGFactory.js';

class LinesCuboid extends Lines {
  constructor(x = 1.0, y = 1.0, z = 1.0, baseZAtZero=false) {
    super();

    this.__x = this.addParameter(new NumberParameter('x', x));
    this.__x.valueChanged.connect(this.__resize.bind(this));
    this.__y = this.addParameter(new NumberParameter('y', y));
    this.__y.valueChanged.connect(this.__resize.bind(this));
    this.__z = this.addParameter(new NumberParameter('z', z));
    this.__z.valueChanged.connect(this.__resize.bind(this));
    this.__baseZAtZero = this.addParameter(new NumberParameter('BaseZAtZero', baseZAtZero));
    this.__baseZAtZero.valueChanged.connect(this.__rebuild.bind(this));
    this.__rebuild();
  }

  __rebuild() {
    this.setNumVertices(8);
    this.setNumSegments(12);
    this.setSegment(0, 0, 1);
    this.setSegment(1, 1, 2);
    this.setSegment(2, 2, 3);
    this.setSegment(3, 3, 0);

    this.setSegment(4, 4, 5);
    this.setSegment(5, 5, 6);
    this.setSegment(6, 6, 7);
    this.setSegment(7, 7, 4);

    this.setSegment(8, 0, 4);
    this.setSegment(9, 1, 5);
    this.setSegment(10, 2, 6);
    this.setSegment(11, 3, 7);
    this.__resize(-1);
    this.geomDataTopologyChanged.emit();
  }

  __resize(mode) {
    const x = this.__x.getValue();
    const y = this.__y.getValue();
    const z = this.__z.getValue();
    const baseZAtZero = this.__baseZAtZero.getValue();

    let zoff = 0.5;
    if(baseZAtZero)
      zoff = 1.0
    this.getVertex(0).set( 0.5 * x, -0.5 * y, zoff * z);
    this.getVertex(1).set( 0.5 * x,  0.5 * y, zoff * z);
    this.getVertex(2).set(-0.5 * x,  0.5 * y, zoff * z);
    this.getVertex(3).set(-0.5 * x, -0.5 * y, zoff * z);

    zoff = -0.5;
    if(baseZAtZero)
      zoff = 0.0
    this.getVertex(4).set( 0.5 * x, -0.5 * y, zoff * z);
    this.getVertex(5).set( 0.5 * x,  0.5 * y, zoff * z);
    this.getVertex(6).set(-0.5 * x,  0.5 * y, zoff * z);
    this.getVertex(7).set(-0.5 * x, -0.5 * y, zoff * z);


    this.setBoundingBoxDirty();
    if(mode != -1)
      this.geomDataChanged.emit();
  }


  toJSON() {
    let json = super.toJSON();
    json['size'] = this.__size;
    return json
  }
};

sgFactory.registerClass('LinesCuboid', LinesCuboid);
export {
  LinesCuboid
};