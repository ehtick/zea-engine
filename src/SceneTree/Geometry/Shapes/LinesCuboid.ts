import { NumberParameter } from '../../Parameters/NumberParameter'
import { Registry } from '../../../Registry'
import { ProceduralLines } from './ProceduralLines'
import { BooleanParameter } from '../../../SceneTree/Parameters'
import { Vec3Attribute } from '../Vec3Attribute'

/**
 * A class for generating a lines cuboid shape(Without faces).
 *
 * **Parameters**
 * * **X(`NumberParameter`):** Length of the line cuboid along the `X` axis
 * * **Y(`NumberParameter`):** Length of the line cuboid along the `Y` axis
 * * **Z(`NumberParameter`):** Length of the line cuboid along the `Z` axis
 * * **BaseZAtZero(`NumberParameter`):** Property to start or not `Z` axis from position `0.
 *
 * @extends {ProceduralLines}
 */
class LinesCuboid extends ProceduralLines {
  /**
   * @member {BooleanParameter} baseZAtZeroParam - Property to start or not `Z` axis from position `0.
   */
  baseZAtZeroParam: BooleanParameter

  /**
   * @member {NumberParameter} sizeXParam - Length of the line cuboid along the `X` axis
   */
  sizeXParam: NumberParameter

  /**
   * @member {NumberParameter} sizeYParam - Length of the line cuboid along the `Y` axis
   */
  sizeYParam: NumberParameter

  /**
   * @member {NumberParameter} sizeZParam - Length of the line cuboid along the `Z` axis
   */
  sizeZParam: NumberParameter

  /**
   * Create a lines cuboid.
   * @param {number} x - The length of the line cuboid along the X axis.
   * @param {number} y - The length of the line cuboid along the Y axis.
   * @param {number} z - The length of the line cuboid along the Z axis.
   * @param {boolean} baseZAtZero - The baseZAtZero value.
   */
  constructor(x = 1.0, y = 1.0, z = 1.0, baseZAtZero = false) {
    super()

    this.sizeXParam = this.addParameter(new NumberParameter('X', x)) as NumberParameter
    this.sizeYParam = this.addParameter(new NumberParameter('Y', y)) as NumberParameter
    this.sizeZParam = this.addParameter(new NumberParameter('Z', z)) as NumberParameter

    this.baseZAtZeroParam = this.addParameter(new BooleanParameter('BaseZAtZero', baseZAtZero)) as BooleanParameter
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    this.setNumVertices(8)
    this.setNumSegments(12)
    this.setSegmentVertexIndices(0, 0, 1)
    this.setSegmentVertexIndices(1, 1, 2)
    this.setSegmentVertexIndices(2, 2, 3)
    this.setSegmentVertexIndices(3, 3, 0)

    this.setSegmentVertexIndices(4, 4, 5)
    this.setSegmentVertexIndices(5, 5, 6)
    this.setSegmentVertexIndices(6, 6, 7)
    this.setSegmentVertexIndices(7, 7, 4)

    this.setSegmentVertexIndices(8, 0, 4)
    this.setSegmentVertexIndices(9, 1, 5)
    this.setSegmentVertexIndices(10, 2, 6)
    this.setSegmentVertexIndices(11, 3, 7)
    this.resize()
  }

  /**
   * The resize method.
   *
   * @private
   */
  resize(): void {
    const x = this.sizeXParam.value || 1.0
    const y = this.sizeYParam.value || 1.0
    const z = this.sizeZParam.value || 1.0
    const baseZAtZero = this.baseZAtZeroParam.value

    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    if (positions) {
      let zoff = 0.5
      if (baseZAtZero) zoff = 1.0
      positions.getValueRef(0).set(0.5 * x, -0.5 * y, zoff * z)
      positions.getValueRef(1).set(0.5 * x, 0.5 * y, zoff * z)
      positions.getValueRef(2).set(-0.5 * x, 0.5 * y, zoff * z)
      positions.getValueRef(3).set(-0.5 * x, -0.5 * y, zoff * z)

      zoff = -0.5
      if (baseZAtZero) zoff = 0.0
      positions.getValueRef(4).set(0.5 * x, -0.5 * y, zoff * z)
      positions.getValueRef(5).set(0.5 * x, 0.5 * y, zoff * z)
      positions.getValueRef(6).set(-0.5 * x, 0.5 * y, zoff * z)
      positions.getValueRef(7).set(-0.5 * x, -0.5 * y, zoff * z)
    }
  }
}

Registry.register('LinesCuboid', LinesCuboid)
export { LinesCuboid }
