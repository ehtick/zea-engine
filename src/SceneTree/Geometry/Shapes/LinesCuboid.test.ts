import { LinesCuboid } from './LinesCuboid'

describe('LinesCuboid', () => {
  it('tests default parameters', () => {
    const linesCuboid = new LinesCuboid()

    expect(linesCuboid.sizeXParam.value).toBe(1)
    expect(linesCuboid.sizeYParam.value).toBe(1)
    expect(linesCuboid.sizeZParam.value).toBe(1)
    expect(linesCuboid.baseZAtZeroParam.value).toBeFalsy()
  })

  it('updates parameters', () => {
    const linesCuboid = new LinesCuboid()
    linesCuboid.sizeXParam.value = 3
    linesCuboid.sizeYParam.value = 3
    linesCuboid.sizeZParam.value = 3
    linesCuboid.baseZAtZeroParam.value = true

    expect(linesCuboid.sizeXParam.value).toBe(3)
    expect(linesCuboid.sizeYParam.value).toBe(3)
    expect(linesCuboid.sizeZParam.value).toBe(3)
    expect(linesCuboid.baseZAtZeroParam.value).toBeTruthy()
  })

  // There's an issue with flags, that are preventing parameters to be exported.
  it('saves to JSON (serialization).', () => {
    const linesCuboid = new LinesCuboid(2, 2, 2, true)
    const outputJSON = linesCuboid.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const linesCuboid = new LinesCuboid()
    const inputJSON = {
      params: { X: { value: 2 }, Y: { value: 2 }, Z: { value: 2 }, BaseZAtZero: { value: true } },
      type: 'LinesCuboid',
      vertexAttributes: {}
    }
    linesCuboid.fromJSON(inputJSON)

    const newLinesCuboid = new LinesCuboid(2, 2, 2, true)
    expect(linesCuboid.toJSON()).toEqual(newLinesCuboid.toJSON())
  })
})
