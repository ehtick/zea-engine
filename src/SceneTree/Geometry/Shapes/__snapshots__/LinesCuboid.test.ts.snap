// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinesCuboid saves to JSON (serialization). 1`] = `
Object {
  "params": Object {
    "BaseZAtZero": Object {
      "value": true,
    },
    "X": Object {
      "value": 2,
    },
    "Y": Object {
      "value": 2,
    },
    "Z": Object {
      "value": 2,
    },
  },
  "type": "LinesCuboid",
  "vertexAttributes": Object {},
}
`;
